File: Sources/zigclient/private/src/Dom/Color.zig
const std = @import("std");
const Self = @This();

r: u8,
g: u8,
b: u8,

pub inline fn darkened(self: Self, comptime strength: f64) Self {
    const strength_c = comptime (1 - strength);

    return Self{
        .r = @floor(self.r * strength_c),
        .g = @floor(self.g * strength_c),
        .b = @floor(self.b * strength_c),
    };
}

pub inline fn lightened(self: Self, comptime strength: f64) Self {
    const strength_a = comptime (1 + strength);

    return Self{
        .r = @min(255, @floor(self.r * strength_a)),
        .g = @min(255, @floor(self.g * strength_a)),
        .b = @min(255, @floor(self.b * strength_a)),
    };
}

pub fn fromHex(comptime hex: *const [6:0]u8) Self {
    comptime {
        const r = std.fmt.parseInt(u8, hex[0..2], 16) catch unreachable;
        const g = std.fmt.parseInt(u8, hex[2..4], 16) catch unreachable;
        const b = std.fmt.parseInt(u8, hex[4..6], 16) catch unreachable;

        return Self{ .r = r, .g = g, .b = b };
    }
}

File: Sources/LegacyClient/Public/Sources/Utils/Color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lightened = exports.darkened = exports.DARKENED_BASE = void 0;
const Memoize_1 = require("./Memoize");
// #00000030
exports.DARKENED_BASE = 0.1875;
/**
 * Darkens colour.
 *
 * @param color - Color code
 * @param strength - Strenth
 */
exports.darkened = (0, Memoize_1.memo)((color, strength) => {
    let r = parseInt(color.slice(1, 3), 16);
    let g = parseInt(color.slice(3, 5), 16);
    let b = parseInt(color.slice(5, 7), 16);
    r = Math.floor(r * (1 - strength));
    g = Math.floor(g * (1 - strength));
    b = Math.floor(b * (1 - strength));
    return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
});
/**
 * ligtens colour.
 *
 * @param color - Color code
 * @param strength - Strenth
 */
exports.lightened = (0, Memoize_1.memo)((color, strength) => {
    let r = parseInt(color.slice(1, 3), 16);
    let g = parseInt(color.slice(3, 5), 16);
    let b = parseInt(color.slice(5, 7), 16);
    r = Math.min(255, Math.floor(r * (1 + strength)));
    g = Math.min(255, Math.floor(g * (1 + strength)));
    b = Math.min(255, Math.floor(b * (1 + strength)));
    return `#${r.toString(16).padStart(2, "0")}${g.toString(16).padStart(2, "0")}${b.toString(16).padStart(2, "0")}`;
});
